### PR TITLE
[6.x] Core/Player: Autoattack/melee spells facing check

### PR DESCRIPTION
**Pull request by P-Kito**
**Changes proposed**:
- In 6.x players can perform auto attacks and melee spells even when facing away from the opponent.
- Can provide video from retail, checked yesterday.

**Target branch(es)**: 6x

**Issues addressed**: ----

**Tests performed**: Builds, Tested Ingame

**Known issues and TODO list**:

### CODE REVIEW

**Comment by Nawuko** [comment]
I know this sounds stupid as fuck. But @P-Kito is completly right. You  can attack targets facing in a complete different direction. Confirmed from retail :+1:

**Comment by Nawuko** [comment]
Ok sorry for the double post. This is not quite right. Autoattacks only work if you are in the model radius (the red circle on the bottom  ... idk what its called) of the npcs. Autoattacks still dosent work even if you are in the meele range of the npc.

**Comment by Ascathor** [comment]
So if not within boundaryreach, you still need to face the npc ?

**Comment by Nawuko** [comment]
Exactly. But this only is the case on some creatures not all. So more testing is needed.

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -1130,11 +1133,24 @@ void Player::Update(uint32 p_time)
...
+                            m_swingErrorMsg = 2;
+                        }
+                    }
+                }
                 else
                 {
                     m_swingErrorMsg = 0;                    // reset swing error state
 
-                    // prevent base and off attack in same time, delay attack at 0.2 sec
+                                                            // prevent base and off attack in same time, delay attack at 0.2 sec
```
Comment (P-Kito) [inline src/server/game/Entities/Player/Player.cpp]: ups :p

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -1119,6 +1119,9 @@ void Player::Update(uint32 p_time)
             // default combat reach 10
             /// @todo add weapon, skill check
 
+            Position victimPos = victim->GetPosition();
```
Comment (joschiwald) [inline src/server/game/Entities/Player/Player.cpp]: you should use a const ref

File: src/server/game/Entities/Player/Player.cpp [hunk 3]
```diff
@@ -1119,6 +1119,9 @@ void Player::Update(uint32 p_time)
...
+            float victimBoundaryRadius = victim->GetBoundaryRadius();
```
Comment (tkrokli) [inline src/server/game/Entities/Player/Player.cpp]: Travis compile error : `no member named 'GetBoundaryRadius' in 'Unit'`

**Comment by P-Kito** [comment]
Nice :+1: Working like on retail now. I will also add meele spells into this PR.

**Comment by tkrokli** [comment]
Couple of suggestions; your PR title : **[6.x] Core/player: Autoattack/melee spells facing check**
***
**Issues addressed:** 
Players can perform auto attacks and melee spells even when facing away from the opponent.

File: src/server/game/Entities/Unit/Unit.cpp [hunk 4]
```diff
@@ -520,6 +520,16 @@ bool Unit::IsWithinMeleeRange(const Unit* obj, float dist) const
...
 }
 
+bool Unit::IsWithinBoundaryRadius(const Unit* obj) const
+{
+    if (!obj || !IsInMap(obj) || !IsInPhase(obj))
+        return false;
+
+    float objBoundaryRadius = std::max(obj->GetBoundaryRadius(), MIN_MELEE_REACH);
+
+    return IsWithinBox(*obj, objBoundaryRadius, objBoundaryRadius, objBoundaryRadius);
```
Comment (Shauren) [inline src/server/game/Entities/Unit/Unit.cpp]: This isnt a good call - IsWithinBox checks a square area, not a circle

Comment (Nawuko) [inline src/server/game/Entities/Unit/Unit.cpp]: See, even the master confirms that. I told you P-Kito!

Comment (P-Kito) [inline src/server/game/Entities/Unit/Unit.cpp]: And I told you I will change it

**Comment by EPS1L0N** [comment]
also should be applied to cone effects

**Comment by Aokromes** [comment]
@P-Kito https://github.com/TrinityCore/TrinityCore/pull/16885#issuecomment-204744091

**Comment by P-Kito** [comment]
Huh, what doesnt build here

**Comment by tkrokli** [comment]
```
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Entities/Object/Position.cpp:135:31: fatal error: 
      no member named 'sqrtf' in namespace 'std'; did you mean simply 'sqrtf'?
    float vectorCenterPoint = std::sqrtf(std::powf(SphereDistX, 2) + std...
                              ^~~~~~~~~~
                              sqrtf
```

**Comment by P-Kito** [comment]
Can someone test this on WoD Core? Just stand behind enemy but in his hitbox (neck on neck)and auto attack + use cone of cold for exmaple

**Comment by Aokromes** [comment]
There is any reason to don't merge this?

File: src/server/game/Entities/Object/Position.cpp [hunk 5]
```diff
@@ -126,6 +126,20 @@ bool Position::IsWithinBox(const Position& center, float xradius, float yradius,
     return true;
 }
 
+bool Position::IsWithinSphere(const Position& center, float radius) const
```
Comment (Shauren) [inline src/server/game/Entities/Object/Position.cpp]: this is exactly what Position::IsInDist does

Comment (P-Kito) [inline src/server/game/Entities/Object/Position.cpp]: Oh shit :D

File: src/server/game/Spells/Spell.cpp [hunk 6]
```diff
@@ -7597,6 +7598,9 @@ WorldObjectSpellConeTargetCheck::WorldObjectSpellConeTargetCheck(float coneAngle
 
 bool WorldObjectSpellConeTargetCheck::operator()(WorldObject* target)
 {
+    if (_caster->GetTypeId() == TYPEID_PLAYER && target->IsInDist(_caster, _caster->GetBoundaryRadius()))
```
Comment (Shauren) [inline src/server/game/Spells/Spell.cpp]: _caster->GetBoundaryRadius()
caster or target hitbox?

Comment (P-Kito) [inline src/server/game/Spells/Spell.cpp]: If the target is within casters boundray radius, the cone effect will hit the target. (Even if the target is behind, but in boundary radius)

**Comment by P-Kito** [comment]
No autoattacks or melee spells, because too far away to do it with not facing:
http://i.imgur.com/pvZOdKB.png

Autoattacks and melee spells even while behind target/not facing target:
http://i.imgur.com/YCzVo8e.png

Autoattacks and melee spells with normal range calculation:
http://i.imgur.com/H3BnPaI.png
This cannot be done when not facing (not in boundary radius):
http://i.imgur.com/YcRXTjE.png

Cone effect tests coming in a minute. (slow compile computer)

**Comment by P-Kito** [comment]
Working as intended :) Tested ingame

**Comment by Shauren** [comment]
86275a397a478a30dad7e00af73029e2b15d4668

**Pull request closed without merging**